src/Preflight/Preflight.php

```diff
@@ -341,47 +339,21 @@ public function preflight($argv): bool
     }
 
     /**
-     * Find the site the user selected based on --root or cwd. If neither of
-     * those result in a site, then we will fall back to the vendor path.
+     * Find the Drupal root of the preferred Drupal site (the one
+     * that shares the `vendor` directory with Drush).
      */
-    protected function findSelectedSite()
+    protected function preferredSite()
     {
-        // TODO: If we want to support ONLY site-local Drush (which is
-        // DIFFERENT than --local), then skip the call to `$preflightArgs->selectedSite`
-        // and just assign `false` to $selectedRoot.
-
-        // Try two approaches.
-        $selectedRoot = $this->preflightArgs->selectedSite($this->environment->cwd());
-        $fallBackPath = $this->preflightArgs->selectedSite($this->config()->get('runtime.drush-script'));
-        return $this->setSelectedSite($selectedRoot, $fallBackPath);
-    }
+        $projectRoot = dirname($this->environment->vendorPath());
```

Comment:
Composer has a configurable location for vendor https://getcomposer.org/doc/06-config.md#vendor-dir

Reply:
There is no API to determine the project root from the vendor directory. Two options:

1. `$a = include $vendor/composer/installed.php; $project_root = $a['root']['installed_path'];`
2. Walk up from `$vendor` until we find a composer.json file.

The problem with 1. is that it breaks if Composer changes its internal implementation of what it writes to the vendor directory; installed.php is not guaranteed to exist, and not guaranteed to have the same contents.

The problem with 2. is that there's nothing preventing a user from putting `vendor` outside the project root, or deep inside a directory that might encounter some other composer.json before it gets to the root.

It's pretty rare for Drupal sites to relocate the vendor directory at all; maybe option 2 is good enough?

Reply:
Maybe I'll try option `1` and fall back to `2` if $installed.php doesn't return recognizable data. That would mean that the code path in `2` would be seldom-exercised, though. We could cover it in a tests, though.